src/types/arrow.rs

```diff
@@ -60,9 +66,11 @@ impl Arrow {
 
     /// Create a unification arrow for a fresh `iden` combinator
     pub fn for_iden<J: Jet>(context: &mut Context<J>) -> Self {
-        // We cheat and make the source and target of iden literally be the same
-        // variable, rather than unifying them. This is slightly more efficient
-        // and shouldn't result in any user-visible consequences.
+        // Throughout this module, when two types are the same, we reuse a
+        // pointer to them rather than creating distinct types and unifying
+        // them. This theoretically could lead to more confusing errors for
+        // the user during type inference, but in practice type inference
+        // is completely opaque and there's no harm in making it moreso.
```

Comment:
We have to experiment and see if this becomes an issue for users

Reply:
Agreed. But I'm pretty confident it won't be an issue ... the internal state of type inference is pretty useless no matter how much we try to avoid shortcuts.